libs/ml/examples/word2vec_distributed_learning/main.cpp

```diff
@@ -0,0 +1,185 @@
+using namespace fetch::ml::ops;
+using namespace fetch::ml::layers;
+using namespace fetch::ml;
+using namespace fetch::ml::dataloaders;
+
+using DataType         = double;
+using TensorType       = fetch::math::Tensor<DataType>;
+using VectorTensorType = std::vector<TensorType>;
+using SizeType         = typename TensorType::SizeType;
+
+// std::string ReadFile(std::string const &path)
+//{
+//  std::ifstream t(path);
+//  return std::string((std::istreambuf_iterator<char>(t)), std::istreambuf_iterator<char>());
+//}
+
+int main(int ac, char **av)
+{
+  if (ac < 2)
+  {
+    std::cout << "Usage : " << av[0] << " PATH/TO/text8" << std::endl;
+    return 1;
+  }
+
+  std::string train_file = av[1];
+
+  std::shared_ptr<Coordinator> coordinator =
+      std::make_shared<Coordinator>(SYNCHRONIZATION_MODE, NUMBER_OF_ITERATIONS);
+
+  std::cout << "FETCH Distributed Word2vec Demo -- Asynchronous" << std::endl;
+  std::srand((unsigned int)std::time(nullptr));
+
+  TrainingParams<TensorType> tp;
+
+  // calc the true starting learning rate
+  tp.starting_learning_rate =
+      static_cast<DataType>(tp.batch_size) * tp.starting_learning_rate_per_sample;
+  tp.ending_learning_rate =
+      static_cast<DataType>(tp.batch_size) * tp.ending_learning_rate_per_sample;
+  tp.learning_rate_param.starting_learning_rate = tp.starting_learning_rate;
+  tp.learning_rate_param.ending_learning_rate   = tp.ending_learning_rate;
+
+  //  GraphW2VLoader<DataType> data_loader(tp.window_size, tp.negative_sample_size, tp.freq_thresh,
+  //                                       tp.max_word_count);
+  // set up dataloader
+  /// DATA LOADING ///
+  //  std::string vocab_file = "/tmp/vocab.txt";
+  //  data_loader.BuildVocab({ReadFile(train_file)}, tp.min_count);
+  //  data_loader.SaveVocab(vocab_file);
+
+  std::vector<std::shared_ptr<TrainingClient<TensorType>>> clients(NUMBER_OF_CLIENTS);
+  for (unsigned int i(0); i < NUMBER_OF_CLIENTS; ++i)
+  {
+    // Instanciate NUMBER_OF_CLIENTS clients
+    clients[i] = std::make_shared<Word2VecClient<TensorType>>(std::to_string(i), tp, train_file,
+                                                              BATCH_SIZE, NUMBER_OF_PEERS);
+  }
+
+  for (unsigned int i(0); i < NUMBER_OF_CLIENTS; ++i)
+  {
+    // Give every client the full list of other clients
+    clients[i]->AddPeers(clients);
+
+    // Give each client pointer to coordinator
+    clients[i]->SetCoordinator(coordinator);
+  }
+
+  // Main loop
+  for (unsigned int it(0); it < NUMBER_OF_ROUNDS; ++it)
+  {
+
+    // Start all clients
+    coordinator->Reset();
+    std::cout << "================= ROUND : " << it << " =================" << std::endl;
+    std::list<std::thread> threads;
+    for (auto &c : clients)
+    {
+      threads.emplace_back([&c] { c->MainLoop(); });
+    }
+
+    // Wait for everyone to finish
+    for (auto &t : threads)
+    {
+      t.join();
+    }
+
+    if (coordinator->GetMode() == CoordinatorMode::ASYNCHRONOUS)
+    {
+      break;
```

Comment:
shouldn't this be continue? Doesn't this break out of the for loop resulting in only 1 round?